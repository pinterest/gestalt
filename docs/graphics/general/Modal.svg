<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><path fill="#fff" d="M21 46c0-8.837 7.163-16 16-16h238c8.837 0 16 7.163 16 16v88c0 8.837-7.163 16-16 16H37c-8.837 0-16-7.163-16-16V46Z"/><path fill="#005C62" fill-rule="evenodd" d="M275 32H37c-7.732 0-14 6.268-14 14v88c0 7.732 6.268 14 14 14h238c7.732 0 14-6.268 14-14V46c0-7.732-6.268-14-14-14ZM37 30c-8.837 0-16 7.163-16 16v88c0 8.837 7.163 16 16 16h238c8.837 0 16-7.163 16-16V46c0-8.837-7.163-16-16-16H37Z" clip-rule="evenodd"/><path fill="#005C62" d="M231.5 128c0-6.627 5.373-12 12-12H263c6.627 0 12 5.373 12 12s-5.373 12-12 12h-19.5c-6.627 0-12-5.373-12-12Z"/><path fill="#fff" d="M242.531 130.637c1.203 0 1.949-.621 1.949-1.614 0-.726-.542-1.285-1.289-1.343v-.071a1.218 1.218 0 0 0 1.02-1.203c0-.871-.652-1.406-1.746-1.406H240v5.637h2.531Zm-1.351-4.762h.984c.563 0 .887.273.887.734 0 .469-.348.735-.985.735h-.886v-1.469Zm0 3.883v-1.637h1.023c.695 0 1.074.281 1.074.809 0 .539-.367.828-1.047.828h-1.05Zm8.152-3.422h-1.137v2.484c0 .59-.304.953-.871.953-.523 0-.793-.308-.793-.918v-2.519h-1.136v2.801c0 1 .578 1.59 1.507 1.59.653 0 1.055-.282 1.258-.782h.07v.692h1.102v-4.301Zm1.465-1.027v1.066h-.672v.863h.672v2.239c0 .835.414 1.175 1.461 1.175.219 0 .43-.023.57-.05v-.84c-.109.011-.187.019-.336.019-.387 0-.558-.179-.558-.566v-1.977h.894v-.863h-.894v-1.066h-1.137Zm3.344 0v1.066h-.672v.863h.672v2.239c0 .835.414 1.175 1.461 1.175.218 0 .429-.023.57-.05v-.84c-.11.011-.188.019-.336.019-.387 0-.559-.179-.559-.566v-1.977h.895v-.863h-.895v-1.066h-1.136Zm4.859 5.418c1.324 0 2.121-.836 2.121-2.243 0-1.394-.809-2.242-2.121-2.242-1.312 0-2.121.852-2.121 2.242 0 1.403.797 2.243 2.121 2.243Zm0-.899c-.609 0-.953-.492-.953-1.344 0-.843.348-1.339.953-1.339.602 0 .953.496.953 1.339 0 .852-.348 1.344-.953 1.344Zm3.051.809h1.137v-2.489c0-.582.324-.957.851-.957.539 0 .793.317.793.922v2.524h1.137v-2.782c0-1.027-.516-1.613-1.473-1.613-.637 0-1.074.301-1.273.785h-.071v-.691h-1.101v4.301Z"/><path fill="#009990" d="M184 128c0-6.627 5.373-12 12-12h19.5c6.627 0 12 5.373 12 12s-5.373 12-12 12H196c-6.627 0-12-5.373-12-12Z"/><path fill="#fff" d="M195.031 130.637c1.203 0 1.949-.621 1.949-1.614 0-.726-.542-1.285-1.289-1.343v-.071a1.218 1.218 0 0 0 1.02-1.203c0-.871-.652-1.406-1.746-1.406H192.5v5.637h2.531Zm-1.351-4.762h.984c.563 0 .887.273.887.734 0 .469-.348.735-.985.735h-.886v-1.469Zm0 3.883v-1.637h1.023c.695 0 1.074.281 1.074.809 0 .539-.367.828-1.047.828h-1.05Zm8.152-3.422h-1.137v2.484c0 .59-.304.953-.871.953-.523 0-.793-.308-.793-.918v-2.519h-1.136v2.801c0 1 .578 1.59 1.507 1.59.653 0 1.055-.282 1.258-.782h.07v.692h1.102v-4.301Zm1.465-1.027v1.066h-.672v.863h.672v2.239c0 .835.414 1.175 1.461 1.175.219 0 .43-.023.57-.05v-.84c-.109.011-.187.019-.336.019-.387 0-.558-.179-.558-.566v-1.977h.894v-.863h-.894v-1.066h-1.137Zm3.344 0v1.066h-.672v.863h.672v2.239c0 .835.414 1.175 1.461 1.175.218 0 .429-.023.57-.05v-.84c-.11.011-.188.019-.336.019-.387 0-.559-.179-.559-.566v-1.977h.895v-.863h-.895v-1.066h-1.136Zm4.859 5.418c1.324 0 2.121-.836 2.121-2.243 0-1.394-.809-2.242-2.121-2.242-1.312 0-2.121.852-2.121 2.242 0 1.403.797 2.243 2.121 2.243Zm0-.899c-.609 0-.953-.492-.953-1.344 0-.843.348-1.339.953-1.339.602 0 .953.496.953 1.339 0 .852-.348 1.344-.953 1.344Zm3.051.809h1.137v-2.489c0-.582.324-.957.851-.957.539 0 .793.317.793.922v2.524h1.137v-2.782c0-1.027-.516-1.613-1.473-1.613-.637 0-1.074.301-1.273.785h-.071v-.691h-1.101v4.301Z"/><rect width="130" height="12" x="37" y="46" fill="#009990" fill-opacity=".5" rx="6"/><rect width="238" height="8" x="37" y="66" fill="#009990" fill-opacity=".5" rx="4"/><rect width="166" height="8" x="37" y="82" fill="#009990" fill-opacity=".5" rx="4"/></svg>