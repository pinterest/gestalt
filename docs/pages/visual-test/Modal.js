// @flow strict
import { type Node as ReactNode } from 'react';
import { Box, Button, ColorSchemeProvider, Flex, Modal, Text } from 'gestalt';

export default function Snapshot(): ReactNode {
  return (
    <ColorSchemeProvider colorScheme="light">
      <Box color="default" display="inlineBlock" height={360} padding={1} width={360}>
        <Modal
          accessibilityModalLabel="Label"
          align="start"
          footer={
            <Flex alignItems="center" gap={2} justifyContent="end">
              <Button onClick={() => {}} text="Cancel" />
              <Button color="red" text="Resume" />
            </Flex>
          }
          heading="Heading"
          onDismiss={() => {}}
          size="sm"
          subHeading="Subheading"
        >
          <Text>Text</Text>
        </Modal>
      </Box>
    </ColorSchemeProvider>
  );
}
