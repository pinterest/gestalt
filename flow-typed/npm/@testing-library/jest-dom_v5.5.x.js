// flow-typed signature: 5291c1af24f46ab1b593d2626cfbb310
// flow-typed version: 7ad57fc095/@testing-library/jest-dom_v5.5.x/flow_>=v0.104.x

declare module '@testing-library/jest-dom' {
  declare type JestMatcherResult = {
    message?: string | (() => string),
    pass: boolean,
    ...
  };

  declare type Result = JestMatcherResult | Promise<JestMatcherResult>;

  declare module.exports: {|
    /**
     * @deprecated
     */
    toBeInTheDOM(container?: HTMLElement): Result,

    toBeInTheDocument(): Result,
    toBeVisible(): Result,
    toBeEmpty(): Result,
    toBeDisabled(): Result,
    toBeEnabled(): Result,
    toBeInvalid(): Result,
    toBeRequired(): Result,
    toBeValid(): Result,
    toContainElement(element: HTMLElement | null): Result,
    toContainHTML(htmlText: string): Result,
    toHaveAttribute(attr: string, value?: any): Result,
    toHaveClass(...classNames: string[]): Result,
    toHaveFocus(): Result,
    toHaveFormValues(expectedValues: { [name: string]: any, ... }): Result,
    toHaveStyle(css: string | { [name: string]: any, ... }): Result,
    toHaveTextContent(
      text: string | RegExp,
      options?: {| normalizeWhitespace: boolean |}
    ): Result,
    toHaveValue(value?: string | string[] | number): Result,
    toHaveDisplayValue(value: string | string[]): Result,
    toBeChecked(): Result,
  |};
}

declare module '@testing-library/jest-dom/extend-expect' {
  declare module.exports: any;
}
