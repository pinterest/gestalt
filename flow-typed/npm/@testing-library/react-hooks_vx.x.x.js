// flow-typed signature: a58b5bd0827f6c3918d84b9e8a8a52ca
// flow-typed version: <<STUB>>/@testing-library/react-hooks_v^3.4.1/flow_v0.127.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@testing-library/react-hooks'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@testing-library/react-hooks' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@testing-library/react-hooks/dont-cleanup-after-each' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/asyncUtils' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/cleanup' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/flush-microtasks' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/pure' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/pure' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/asyncUtils' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/cleanup' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/flush-microtasks' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/pure' {
  declare module.exports: any;
}

// Filename aliases
declare module '@testing-library/react-hooks/dont-cleanup-after-each.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/dont-cleanup-after-each'>;
}
declare module '@testing-library/react-hooks/lib/asyncUtils.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/asyncUtils'>;
}
declare module '@testing-library/react-hooks/lib/cleanup.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/cleanup'>;
}
declare module '@testing-library/react-hooks/lib/flush-microtasks.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/flush-microtasks'>;
}
declare module '@testing-library/react-hooks/lib/index' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib'>;
}
declare module '@testing-library/react-hooks/lib/index.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib'>;
}
declare module '@testing-library/react-hooks/lib/pure.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/pure'>;
}
declare module '@testing-library/react-hooks/pure.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/pure'>;
}
declare module '@testing-library/react-hooks/src/asyncUtils.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/asyncUtils'>;
}
declare module '@testing-library/react-hooks/src/cleanup.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/cleanup'>;
}
declare module '@testing-library/react-hooks/src/flush-microtasks.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/flush-microtasks'>;
}
declare module '@testing-library/react-hooks/src/index' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src'>;
}
declare module '@testing-library/react-hooks/src/index.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src'>;
}
declare module '@testing-library/react-hooks/src/pure.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/pure'>;
}
