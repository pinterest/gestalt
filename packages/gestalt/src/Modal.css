.Backdrop {
  composes: absolute left0 top0 right0 bottom0 from "./Layout.css";
  composes: zoomOut from "./Cursor.css";
  composes: darkGrayBg from "./Colors.css";
  height: 100%;
  opacity: 0.9;
  overflow-y: scroll;
}

.container {
  composes: fixed borderBox flex justifyCenter left0 top0 itemsCenter from "./Layout.css";
  composes: xsCol12 from "./Column.css";
  height: 100%;
}

.wrapper {
  composes: relative overflowHidden flex from "./Layout.css";
  composes: whiteBg from "./Colors.css";
  composes: rounded from "./Borders.css";
  composes: marginLeft4 marginRight4 from "./boxWhitespace.css";
  max-height: calc(100vh - 32px);
  outline: none;
}
