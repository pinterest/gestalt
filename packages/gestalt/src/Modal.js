// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import Box from './Box.js';
import Divider from './Divider.js';
import Heading from './Heading.js';
import IconButton from './IconButton.js';
import StopScrollBehavior from './behaviors/StopScrollBehavior.js';
import TrapFocusBehavior from './behaviors/TrapFocusBehavior.js';
import OutsideEventBehavior from './behaviors/OutsideEventBehavior.js';
import styles from './Modal.css';

type Props = {|
  accessibilityCloseLabel: string,
  accessibilityModalLabel: string,
  children?: React.Node,
  footer?: React.Node,
  heading: string | React.Node,
  onDismiss: () => void,
  role?: 'alertdialog' | 'dialog',
  size?: 'sm' | 'md' | 'lg' | number,
|};

const SIZE_WIDTH_MAP = {
  sm: 414,
  md: 544,
  lg: 804,
};

const ESCAPE_KEY_CODE = 27;

function Backdrop({ children }: { children?: React.Node }) {
  return (
    <React.Fragment>
      <div className={styles.Backdrop} />
      {children}
    </React.Fragment>
  );
}

function Header({
  heading,
  role,
}: {
  heading: string | React.Node,
  role?: 'alertdialog' | 'dialog',
}) {
  if (typeof heading !== 'string') {
    return heading;
  }

  if (role === 'dialog') {
    return (
      <Box
        dangerouslySetInlineStyle={{
          __style: { paddingLeft: 50, paddingRight: 50 },
        }}
        display="flex"
        justifyContent="center"
        paddingY={5}
      >
        <Heading size="xs" accessibilityLevel={1}>
          {heading}
        </Heading>
      </Box>
    );
  }

  return (
    <Box display="flex" padding={4}>
      <Heading size="sm" accessibilityLevel={1}>
        {heading}
      </Heading>
    </Box>
  );
}

export default class Modal extends React.Component<Props> {
  static propTypes = {
    accessibilityCloseLabel: PropTypes.string.isRequired,
    accessibilityModalLabel: PropTypes.string.isRequired,
    children: PropTypes.node,
    footer: PropTypes.node,
    heading: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    onDismiss: PropTypes.func,
    role: PropTypes.oneOf(['alertdialog', 'dialog']),
    size: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.oneOf(['sm', 'md', 'lg']),
    ]),
  };

  componentDidMount() {
    window.addEventListener('keyup', this.handleKeyUp);
  }

  componentWillUnmount() {
    window.removeEventListener('keyup', this.handleKeyUp);
  }

  handleOutsideClick = () => {
    this.props.onDismiss();
  };

  handleCloseClick = () => {
    this.props.onDismiss();
  };

  handleKeyUp = (event: { keyCode: number }) => {
    if (event.keyCode === ESCAPE_KEY_CODE) {
      this.props.onDismiss();
    }
  };

  render() {
    const {
      accessibilityCloseLabel,
      accessibilityModalLabel,
      children,
      footer,
      heading,
      role = 'dialog',
      size = 'sm',
    } = this.props;

    const width = typeof size === 'string' ? SIZE_WIDTH_MAP[size] : size;

    return (
      <StopScrollBehavior>
        <TrapFocusBehavior>
          <div
            aria-label={accessibilityModalLabel}
            className={styles.container}
            role={role}
          >
            <Backdrop>
              <OutsideEventBehavior onClick={this.handleOutsideClick}>
                <div className={styles.wrapper} tabIndex={-1} style={{ width }}>
                  <Box
                    flex="grow"
                    position="relative"
                    display="flex"
                    direction="column"
                    width="100%"
                  >
                    <Box fit>
                      <Header heading={heading} role={role} />
                      {role === 'dialog' && (
                        <React.Fragment>
                          <Box padding={2} position="absolute" top right>
                            <IconButton
                              accessibilityLabel={accessibilityCloseLabel}
                              icon="cancel"
                              onClick={this.handleCloseClick}
                            />
                          </Box>
                          <Divider />
                        </React.Fragment>
                      )}
                    </Box>
                    <Box flex="grow" overflow="auto" position="relative">
                      {children}
                    </Box>
                    <Box fit>
                      {footer && (
                        <Box>
                          {role === 'dialog' && <Divider />}
                          <Box padding={4}>{footer}</Box>
                        </Box>
                      )}
                    </Box>
                  </Box>
                </div>
              </OutsideEventBehavior>
            </Backdrop>
          </div>
        </TrapFocusBehavior>
      </StopScrollBehavior>
    );
  }
}
