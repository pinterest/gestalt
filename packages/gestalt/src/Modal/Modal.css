.overlay {
  height: 100%;
  opacity: 0.9;
  overflow-y: scroll;
}

.container {
  height: 100%;
}

.wrapper {
  max-height: 90%;
  outline: none;
}
