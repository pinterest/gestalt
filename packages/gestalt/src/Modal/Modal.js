// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Box from '../Box/Box';
import Divider from '../Divider/Divider';
import Heading from '../Heading/Heading';
import IconButton from '../IconButton/IconButton';

import styles from './Modal.css';
import borders from '../Borders.css';
import breakpoints from '../breakpoints.json';
import colors from '../Colors.css';
import column from '../Column.css';
import cursor from '../Cursor.css';
import layout from '../Layout.css';
import whitespace from '../Whitespace.css';

const SIZE_WIDTH_MAP = {
  sm: 414,
  md: 544,
  lg: 804,
};

const ESCAPE_KEY_CODE = 27;

type Props = {|
  accessibilityCloseLabel: string,
  accessibilityModalLabel: string,
  children?: React.Node,
  footer?: React.Node,
  heading: string,
  onDismiss: () => void,
  role?: 'alertdialog' | 'dialog',
  size?: 'sm' | 'md' | 'lg',
|};

type State = {
  breakpoint: 'xs' | 'sm' | 'md' | 'lg',
  windowHeight: ?number,
};

export default class Modal extends React.Component<Props, State> {
  state: State = {
    breakpoint: 'xs',
    windowHeight: undefined,
  };

  componentDidMount() {
    document.addEventListener('click', this.handlePageClick);
    window.addEventListener('keydown', this.handleKeyDown);
    window.addEventListener('resize', this.updateBreakpoint);
    document.addEventListener('focus', this.restrictFocus, true);
    this.priorFocus = document.activeElement;
    this.updateBreakpoint();
    if (document.body) {
      document.body.style.overflow = 'hidden'; // Prevents background scrolling
    }
    if (this.modal && this.modal.focus) {
      // Checking this.modal.focus to address a bug in IE11
      // Though the modal exists, this.modal.focus may be null
      // See http://www.mkyong.com/javascript/focus-is-not-working-in-ie-solution/
      this.modal.focus();
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handlePageClick);
    window.removeEventListener('keydown', this.handleKeyDown);
    window.removeEventListener('resize', this.updateBreakpoint);
    document.removeEventListener('focus', this.restrictFocus, true);

    if (document.body) {
      document.body.style.overflow = ''; // Reenables background scrolling
    }
    if (this.priorFocus && this.priorFocus.focus) {
      // Checking this.priorFocus.focus to address a bug in IE11
      this.priorFocus.focus();
    }
  }

  getCurrentBreakpoint = () => {
    let size = 'xs';
    Object.keys(breakpoints).forEach(point => {
      if (window.matchMedia(breakpoints[point]).matches) {
        size = point;
      }
    });
    return size;
  };

  handleClose = () => {
    this.props.onDismiss();
  };

  handlePageClick = (event: Event) => {
    if (
      event.target instanceof Node &&
      this.container &&
      this.container.contains(event.target) &&
      this.modal &&
      !this.modal.contains(event.target)
    ) {
      this.handleClose();
    }
  };

  handleKeyDown = (event: { keyCode: number }) => {
    if (event.keyCode === ESCAPE_KEY_CODE) {
      this.handleClose();
    }
  };

  updateBreakpoint = () => {
    const size = this.getCurrentBreakpoint();
    if (
      size !== this.state.breakpoint ||
      window.innerHeight !== this.state.windowHeight
    ) {
      this.setState({ breakpoint: size, windowHeight: window.innerHeight });
    }
  };

  restrictFocus = (event: Event) => {
    if (
      event.target instanceof Node &&
      this.modal &&
      !this.modal.contains(event.target)
    ) {
      this.modal.focus();
    }
  };

  container: ?HTMLElement;
  modal: ?HTMLElement;
  priorFocus: ?HTMLElement;

  render() {
    const {
      accessibilityCloseLabel,
      accessibilityModalLabel,
      children,
      footer,
      heading,
      role = 'dialog',
      size = 'sm',
    } = this.props;
    const width = SIZE_WIDTH_MAP[size];

    const container = [
      layout.fixed,
      layout.borderBox,
      layout.flex,
      layout.justifyCenter,
      layout.left0,
      layout.top0,
    ];
    const containerClasses =
      this.state.breakpoint === 'xs'
        ? classnames(container, layout.itemsEnd, layout.bottom0, column.xsCol12)
        : classnames(
            container,
            layout.itemsCenter,
            column.xsCol12,
            styles.container
          );

    const wrapper = [layout.fit, layout.relative];
    const wrapperClasses =
      this.state.breakpoint === 'xs'
        ? classnames(wrapper, colors.whiteBg, whitespace.m0, layout.selfEnd)
        : classnames(wrapper, colors.whiteBg, borders.rounded, styles.wrapper);

    const overlay = [
      layout.absolute,
      layout.left0,
      layout.top0,
      cursor.zoomOut,
    ];
    const overlayClasses = classnames(
      overlay,
      styles.overlay,
      colors.darkGrayBg,
      column.xsCol12
    );

    return (
      <div
        aria-label={accessibilityModalLabel}
        className={containerClasses}
        ref={c => {
          this.container = c;
        }}
        role={role}
      >
        <div className={overlayClasses} />
        <div
          className={wrapperClasses}
          ref={c => {
            this.modal = c;
          }}
          tabIndex={-1}
          style={{ width }}
        >
          <Box
            maxHeight="90vh"
            position="relative"
            xs={{ display: 'flexColumn' }}
          >
            <Box fit>
              {role === 'dialog' ? (
                <Box
                  dangerouslySetInlineStyle={{
                    __style: { paddingLeft: 50, paddingRight: 50 },
                  }}
                  display="flex"
                  justifyContent="center"
                  paddingY={5}
                >
                  <Heading size="xs" accessibilityLevel={1}>
                    {heading}
                  </Heading>
                </Box>
              ) : (
                <Box display="flex" padding={4}>
                  <Heading size="lg" accessibilityLevel={1}>
                    {heading}
                  </Heading>
                </Box>
              )}
              {role === 'dialog' && (
                <Box padding={2} position="absolute" top right>
                  <IconButton
                    accessibilityLabel={accessibilityCloseLabel}
                    icon="cancel"
                    onClick={this.handleClose}
                  />
                </Box>
              )}
              {role === 'dialog' && <Divider />}
            </Box>
            <Box flex="grow" overflow="auto" position="relative">
              {children}
            </Box>
            <Box fit>
              {footer && (
                <Box>
                  {role === 'dialog' && <Divider />}
                  <Box padding={4}>{footer}</Box>
                </Box>
              )}
            </Box>
          </Box>
        </div>
      </div>
    );
  }
}

Modal.propTypes = {
  children: PropTypes.node,
  accessibilityCloseLabel: PropTypes.string.isRequired,
  footer: PropTypes.node,
  heading: PropTypes.string.isRequired,
  accessibilityModalLabel: PropTypes.string.isRequired,
  onDismiss: PropTypes.func,
  role: PropTypes.oneOf(['alertdialog', 'dialog']),
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
};
