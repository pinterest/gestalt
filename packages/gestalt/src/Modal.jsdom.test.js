// @flow strict
import React from 'react';
import { render } from '@testing-library/react';
import Modal from './Modal.js';

const mockOnDismiss = jest.fn();

describe('Modal', () => {
  if (typeof document !== 'undefined') {
    test('Modal renders with ref', () => {
      const modalRef = React.createRef();
      const modal = render(
        <Modal accessibilityModalLabel="Test modal" onDismiss={mockOnDismiss} ref={modalRef}>
          Modal content
        </Modal>,
      );

      expect(modal).toMatchSnapshot();
      expect(modalRef.current instanceof HTMLDivElement).toEqual(true);
    });
  }
});
