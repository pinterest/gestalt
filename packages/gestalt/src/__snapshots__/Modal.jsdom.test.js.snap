// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Modal renders with ref 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="overflow: hidden;"
  >
    <div>
      <div>
        <div
          aria-label="Test modal"
          class="container"
          role="dialog"
        >
          <div
            class="backdrop zoomOut"
          />
          <div
            class="wrapper"
            style="width: 540px;"
            tabindex="-1"
          >
            <div
              class="box flexGrow relative xsDirectionColumn xsDisplayFlex"
              style="width: 100%;"
            >
              <div
                class="box flexGrow overflowAuto"
              >
                Modal content
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        aria-label="Test modal"
        class="container"
        role="dialog"
      >
        <div
          class="backdrop zoomOut"
        />
        <div
          class="wrapper"
          style="width: 540px;"
          tabindex="-1"
        >
          <div
            class="box flexGrow relative xsDirectionColumn xsDisplayFlex"
            style="width: 100%;"
          >
            <div
              class="box flexGrow overflowAuto"
            >
              Modal content
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
